File: apps/api/app/research/activity.py
"""Agent, tool and model call records: the step-by-step trace (FR-10).

Three linked layers that together answer "why did the run do that?" without
needing server logs. Written by the worker as it runs the graph (Phase 16,
``app.observability.ledger``) and read back through ``ActivityStore``.
"""

from __future__ import annotations

from datetime import datetime
from typing import Protocol
from uuid import UUID

from app.core.enums import (
    AgentName,
    AgentStatus,
    LlmCallStatus,
    LlmProvider,
    ToolName,
    ToolStatus,
)
from app.research.schemas import ApiModel, RunError


class AgentRunRecord(ApiModel):
    id: UUID
    run_id: UUID
    task_external_id: str | None
    agent_name: AgentName
    iteration: int
    status: AgentStatus
    summary: str
    latency_ms: int | None
    tokens: int
    cost_usd: float
    trace_id: str | None
    span_id: str | None
    error: RunError | None
    started_at: datetime
    completed_at: datetime | None


class ToolCallRecord(ApiModel):
    id: UUID
    agent_run_id: UUID
    tool_name: ToolName
    request_summary: str
    status: ToolStatus
    latency_ms: int
    cache_hit: bool
    retries: int
    result_summary: str
    started_at: datetime


class LlmCallRecord(ApiModel):
    """One model call with the accounting Phase 16 requires on every call."""

    id: UUID
    agent_run_id: UUID | None
    role: AgentName
    provider: LlmProvider
    model: str
    prompt_version: str
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    cost_usd: float
    latency_ms: int
    cache_hit: bool
    status: LlmCallStatus
    created_at: datetime


class ActivityResponse(ApiModel):
    agent_runs: list[AgentRunRecord]
    tool_calls: list[ToolCallRecord]
    llm_calls: list[LlmCallRecord]


class ActivityStore(Protocol):
    """Where the trace is read from.

    The protocol lives beside the DTOs, and the Postgres implementation lives
    in ``app/db/repositories/trace.py`` - the same arrangement as the evidence
    and report stores.
    """

    async def activity_for(self, run_id: UUID) -> ActivityResponse:
        """This run's trace, bounded. Empty for a run that has not started."""
        ...

File: packages/shared-types/src/activity.ts
import type { IsoDateTime, Uuid } from './common';
import type {
  AgentName,
  AgentStatus,
  LlmCallStatus,
  LlmProvider,
  ToolName,
  ToolStatus,
} from './enums';

/** One agent execution. The unit of the step-by-step trace. */
export interface AgentRunRecord {
  id: Uuid;
  run_id: Uuid;
  task_external_id: string | null;
  agent_name: AgentName;
  iteration: number;
  status: AgentStatus;
  /** One-line human summary, e.g. "decomposed into 7 subtasks". */
  summary: string;
  latency_ms: number | null;
  tokens: number;
  cost_usd: number;
  /** OpenTelemetry ids, so a UI row links to a trace in Phase 17. */
  trace_id: string | null;
  span_id: string | null;
  error: { code: string; message: string } | null;
  started_at: IsoDateTime;
  completed_at: IsoDateTime | null;
}

/** One external action taken by an agent: a search, a fetch, an API call. */
export interface ToolCallRecord {
  id: Uuid;
  agent_run_id: Uuid;
  tool_name: ToolName;
  /** Compact display of the request, e.g. the query or the URL. */
  request_summary: string;
  status: ToolStatus;
  latency_ms: number;
  cache_hit: boolean;
  retries: number;
  /** Short outcome, e.g. "12 results" or "403 forbidden". */
  result_summary: string;
  started_at: IsoDateTime;
}

/** One model call, with the accounting Phase 16 requires. */
export interface LlmCallRecord {
  id: Uuid;
  agent_run_id: Uuid | null;
  role: AgentName;
  provider: LlmProvider;
  model: string;
  prompt_version: string;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  cost_usd: number;
  latency_ms: number;
  cache_hit: boolean;
  status: LlmCallStatus;
  created_at: IsoDateTime;
}

/** The full trace behind /research/[id]/activity. */
export interface ActivityResponse {
  agent_runs: AgentRunRecord[];
  tool_calls: ToolCallRecord[];
  llm_calls: LlmCallRecord[];
}
